topdown/object.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func builtinObjectUnion(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	objA, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	objB, err := builtins.ObjectOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	r := mergeWithOverwrite(objA, objB)
+
+	return iter(ast.NewTerm(r))
+}
+
+func builtinObjectRemove(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Expect an object and an array/set of keys
+	obj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Build a set of keys to remove
+	keysToRemove := getObjectKeysParam(operands[1].Value)
+
+	r := ast.NewObject()
+	obj.Foreach(func(key *ast.Term, value *ast.Term) {
+		if !keysToRemove.Contains(key) {
+			r.Insert(key, value)
+		}
+	})
+
+	return iter(ast.NewTerm(r))
+}
+
+func builtinObjectFilter(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Expect an object and an array/set of keys
+	obj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Build a new object from the supplied filter keys
+	keys := getObjectKeysParam(operands[1].Value)
+	filterObj := ast.NewObject()
+	keys.Foreach(func(key *ast.Term) {
+		filterObj.Insert(key, ast.NullTerm())
+	})
+
+	// Actually do the filtering
+	r, err := obj.Filter(filterObj)
+	if err != nil {
+		return err
+	}
```

Comment:
I'd probably leave a short comment mentioning that we could improve this to iterate over the smaller of the two.

Reply:
Not 100% sure I follow, you mean when building the `filterObj` or in the actual filtering?

The implementation for `object.Filter` already does that optimization when doing the actual filtering.

Reply:
Well, since we're using the `ast.Object#Filter` function we have to build the `filterObj` -- the alternative would be to handle the filtering as a special case here.